closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -460,6 +519,68 @@ int decryptFinal(byte[] in, int inOfs, int len,
         return ret;
     }
 
+
+    /**
+     * Performs decryption operation for the last time.
+     *
+     * @param src the input buffer with the data to be decrypted
+     * @param dst the offset in <code>in</code>
+     * @return the number of bytes placed into the <code>out</code> buffer
+     */
+    int decryptFinal(ByteBuffer src, ByteBuffer dst)
+        throws IllegalBlockSizeException, AEADBadTagException,
+        ShortBufferException {
+
+        // Get array from source
+        byte[] src_arr = getBbArray(src);
+        int src_ofs = src.position() + (src.hasArray() ? src.arrayOffset() : 0);
+
+        // Get array from destination
+        byte[] dst_arr = null;
+        if (dst.hasArray()) {
+            dst_arr = dst.array();
+        }
+        else {
+            dst_arr = new byte[dst.capacity()];
+        }
+        int dst_offset = dst.position() + (dst.hasArray() ? dst.arrayOffset() : 0);
+
+        int len = decryptFinal(src_arr, src_ofs, src.remaining(), dst_arr, dst_offset);
+        
+        // Update destination buffer
+        if (dst.hasArray()) {
+            dst.position(dst.position() + len);
+        }
+        else {
+            dst.put(dst_arr, dst.position(), len);       
+        }
+
+        if (dst.position() < dst.limit()) {
+            dst.limit(dst.position());
+        }
+        return len;
+    }
+
+    /**
+     * Gets the byte array behind a buffer.
+     * Tries to use ByteBuffer.array(). If this is not available, the function uses ByteBuffer.get()
+     *
+     * @param src the buffer whose byte[] is needed
+     * @return the byte array with the buffer's content
+     */
+    private byte[] getBbArray(ByteBuffer src) {
+        byte[] arr;
+        if (src.hasArray()) arr = src.array();
+        else {
```

Comment:
Please put body of `if` into a block:
```
        if (src.hasArray()) {
            arr = src.array();
        } else {
```
Likewise, line 579 below.

Reply:
Done